File: tests/sdk/nodejs/test/rds-engines.test.mjs
import { test } from "node:test";
import assert from "node:assert/strict";
import {
  requireReady,
  signedFetch,
  sleep,
  uniquePrefix,
} from "../lib/helpers.mjs";

async function rdsForm(params) {
  const body = new URLSearchParams(params).toString();
  const resp = await signedFetch(
    "rds",
    "POST",
    "/",
    body,
    "application/x-www-form-urlencoded",
  );
  const text = await resp.text();
  return { status: resp.status, body: text };
}

test("RDS MySQL and MariaDB Create Describe Delete soft-skip nested", async (t) => {
  if (!(await requireReady(t))) {
    return;
  }
  const prefix = uniquePrefix();
  for (const engine of ["mysql", "mariadb"]) {
    await t.test(engine, async (t) => {
      let id = `${prefix}-${engine}`.toLowerCase().replace(/_/g, "-");
      if (id.length > 60) {
        id = id.slice(0, 60);
      }
      const create = await rdsForm({
        Action: "CreateDBInstance",
        Version: "2014-10-31",
        DBInstanceIdentifier: id,
        Engine: engine,
        DBInstanceClass: "db.t3.micro",
        MasterUsername: "root",
        MasterUserPassword: "lab-password-1",
        AllocatedStorage: "20",
        DBName: "appdb",
      });
      assert.equal(create.status, 200, create.body);
      assert.match(create.body, new RegExp(id));
      assert.match(create.body, new RegExp(engine));
      t.after(async () => {
        await rdsForm({
          Action: "DeleteDBInstance",
          Version: "2014-10-31",
          DBInstanceIdentifier: id,
        });
      });

      const desc = await rdsForm({
        Action: "DescribeDBInstances",
        Version: "2014-10-31",
        DBInstanceIdentifier: id,
      });
      assert.equal(desc.status, 200, desc.body);
      assert.match(desc.body, new RegExp(id));

      if (process.env.NOCTAXRIS_NESTED === "1") {
        let last = desc.body;
        for (let i = 0; i < 45; i++) {
          const poll = await rdsForm({
            Action: "DescribeDBInstances",
            Version: "2014-10-31",
            DBInstanceIdentifier: id,
          });
          last = poll.body;
          if (last.includes("available")) {
            break;
          }
          if (last.includes("failed")) {
            t.skip(
              `RDS ${engine} nested soft-skip: status failed (nested engine not healthy)`,
            );
            return;
          }
          await sleep(2000);
        }
        if (!last.includes("available")) {
          t.skip(
            `RDS ${engine} nested soft-skip: status not available (healthy DinD required)`,
          );
          return;
        }
      }

      const del = await rdsForm({
        Action: "DeleteDBInstance",
        Version: "2014-10-31",
        DBInstanceIdentifier: id,
      });
      assert.equal(del.status, 200, del.body);
    });
  }
});

File: tests/sdk/go/rds_engines_test.go
package sdk_test

import (
	"net/http"
	"net/url"
	"os"
	"strings"
	"testing"
	"time"
)

func TestRDSMySQLMariaDBCreateDescribeDelete(t *testing.T) {
	requireReady(t)
	prefix := uniquePrefix(t)
	engines := []struct {
		name   string
		engine string
	}{
		{name: "mysql", engine: "mysql"},
		{name: "mariadb", engine: "mariadb"},
	}
	for _, tc := range engines {
		t.Run(tc.name, func(t *testing.T) {
			id := strings.ToLower(strings.ReplaceAll(prefix+"-"+tc.name, "_", "-"))
			if len(id) > 60 {
				id = id[:60]
			}
			createBody := url.Values{
				"Action":               {"CreateDBInstance"},
				"Version":              {"2014-10-31"},
				"DBInstanceIdentifier": {id},
				"Engine":               {tc.engine},
				"DBInstanceClass":      {"db.t3.micro"},
				"MasterUsername":       {"root"},
				"MasterUserPassword":   {"lab-password-1"},
				"AllocatedStorage":     {"20"},
				"DBName":               {"appdb"},
			}.Encode()
			status, body := signedHTTP(t, "rds", http.MethodPost, "/", []byte(createBody), "application/x-www-form-urlencoded")
			if status != http.StatusOK {
				t.Fatalf("CreateDBInstance status=%d body=%s", status, body)
			}
			if !strings.Contains(string(body), id) || !strings.Contains(string(body), tc.engine) {
				t.Fatalf("create missing id/engine: %s", body)
			}
			t.Cleanup(func() {
				delBody := url.Values{
					"Action":               {"DeleteDBInstance"},
					"Version":              {"2014-10-31"},
					"DBInstanceIdentifier": {id},
				}.Encode()
				_, _ = signedHTTP(t, "rds", http.MethodPost, "/", []byte(delBody), "application/x-www-form-urlencoded")
			})

			descBody := url.Values{
				"Action":               {"DescribeDBInstances"},
				"Version":              {"2014-10-31"},
				"DBInstanceIdentifier": {id},
			}.Encode()
			status, body = signedHTTP(t, "rds", http.MethodPost, "/", []byte(descBody), "application/x-www-form-urlencoded")
			if status != http.StatusOK || !strings.Contains(string(body), id) {
				t.Fatalf("DescribeDBInstances status=%d body=%s", status, body)
			}
			xml := string(body)
			if !strings.Contains(xml, "<Engine>"+tc.engine+"</Engine>") && !strings.Contains(xml, tc.engine) {
				t.Fatalf("describe missing engine %s: %s", tc.engine, xml)
			}

			if strings.TrimSpace(os.Getenv("NOCTAXRIS_NESTED")) == "1" {
				deadline := time.Now().Add(90 * time.Second)
				for time.Now().Before(deadline) {
					status, body = signedHTTP(t, "rds", http.MethodPost, "/", []byte(descBody), "application/x-www-form-urlencoded")
					if status == http.StatusOK && strings.Contains(string(body), "available") {
						break
					}
					if status == http.StatusOK && strings.Contains(string(body), "failed") {
						t.Skipf("RDS %s nested soft-skip: status failed (nested engine not healthy)", tc.engine)
					}
					time.Sleep(2 * time.Second)
				}
				if !strings.Contains(string(body), "available") {
					t.Skipf("RDS %s nested soft-skip: status not available (set NOCTAXRIS_NESTED=1 with healthy DinD)", tc.engine)
				}
			}

			delBody := url.Values{
				"Action":               {"DeleteDBInstance"},
				"Version":              {"2014-10-31"},
				"DBInstanceIdentifier": {id},
			}.Encode()
			status, body = signedHTTP(t, "rds", http.MethodPost, "/", []byte(delBody), "application/x-www-form-urlencoded")
			if status != http.StatusOK {
				t.Fatalf("DeleteDBInstance status=%d body=%s", status, body)
			}
		})
	}
}
